Initial commit for JWT Authenticatio&Authorization

enciney opened the pull request:
* JWT authentication integrated to the Swagger page
* debugSettings:IsWebApiEnabled  key introduced , use it for Json friendly Web API
* JwtTokenSettings introduced with Issuer, Audience and Secret
* JWT token validation, authentication and basic authorization is introduced
* [ApiExplorerSettings(IgnoreApi = true)] added ignore old MVC endpoints from swagger

@erenkorkmazemre please use below setup to work on it ; 
* Open appsettings.json and change IsWebApiEnabled  to true
* open _https://localhost:5401/swagger/index.html_ and try to use **api/AuthenticationWeb/Login**
* You can use admin user/ credentials ( or you can try to create new user with Sign Up , it will have _User_ role automatically)
* **api/AuthenticationWeb/Login** this request will return you a JWT token as string
* Now you can copy  it and click the Authorize button popup inside swagger and click Authorize 
![image](https://github.com/user-attachments/assets/e7c2a18d-026b-4ed9-9ca2-11e264fe501f)
* Now you can use MessageWeb/api/MessageWeb/ChatIndex with 1 value in both parameter
![image](https://github.com/user-attachments/assets/83451b31-c350-4d75-8c44-5ac930e6afa8)
* Voooaaalaaahh it will return Json values :)

enciney committed:
Initial commit for JWT Authenticatio&Authorization

erenkorkmazemre commented on EternalBAND.Win/Controllers/WebApi/ErrorWebController.cs:
```diff
@@ -0,0 +1,36 @@
+                case (int)HttpStatusCode.NotFound :
+                    return NotFound("page is not found");
+                case (int)HttpStatusCode.Unauthorized:
+                    return Unauthorized("USer is not logged in for this action");
```
USer

United States with er <3

enciney replied:
no no in german -er postfix means ".. from" for example ;  Hamburger : from Hamburg
USer means from us : bizdensin dostum :D

erenkorkmazemre replied:
OMFG

enciney replied:
controller is deleted and middleware is added

erenkorkmazemre commented on EternalBAND.Win/Controllers/WebApi/AuthentiationWebController.cs:
```diff
@@ -0,0 +1,99 @@
+                {
+                    return Unauthorized("Username or password is incorrect");
+                }
+            }
```
if (!result.Succeeded)
            {
                if (result.IsNotAllowed)
                {
                    return Unauthorized("User is not allowed");
                }

                else if (result.IsLockedOut)
                {
                    return Unauthorized("User account is locked");
                }

                else if (result.RequiresTwoFactor)
                {
                    return Unauthorized("Need 2 FActor Authentication");
                }

                else
                {
                    return Unauthorized("Username or password is incorrect");
                }
            }
            
            
I feel it these lines is written the late nights. Lets discuss it just before we enter Pubg.

enciney replied:
ControllerBase Class has lots of helpful method but I did not discover it honestly.
What we can do is creating dedicated controller or middleware that handle status codes ( We can change in ErrorWebController or convert it to a middleware )

enciney replied:
@erenkorkmazemre  added an extension method HandleSignInFailure to cover signIn problems please take a look I believe it is clearer now

erenkorkmazemre commented on EternalBAND.Api/Options/JwtTokenOptions.cs:
```diff
@@ -0,0 +1,11 @@
+    {
+        public const string JwtOptionKey = "JwtTokenSettings";
+
+        public string Secret { get; set; } = "";
```
public string Secret { get; set; } = "";

Let's talk later about where this private key will be stored.

enciney replied:
there is an option to hold kind of secrets in .net name is .net secret manager
https://learn.microsoft.com/en-us/aspnet/core/security/app-secrets?view=aspnetcore-8.0&tabs=windows

Of course we can implement other logics if you have any suggestions

erenkorkmazemre commented on EternalBAND.Win/Controllers/WebApi/ErrorWebController.cs:
```diff
@@ -0,0 +1,36 @@
+        public ErrorWebController()
+        {
+        }
+
```
public ErrorWebController()
        {
        }
        
is this constructor necessary? In Java If a class has no constructors, the Java compiler automatically provides a default constructor. This is called an implicit default constructor.

enciney replied:
Time to time at the IOC level if you register some class from interface it is giving compiler error but for controller it should be not a problem, I will prune and give a shot.

enciney committed:
Logout and Google sign is added

erenkorkmazemre commented on EternalBAND.Api/Services/AuthenticationService.cs:
```diff
@@ -0,0 +1,62 @@
+        private string CreateToken(Users? user, IEnumerable<string> roles)
+        {
+            var tokenHandler = new JwtSecurityTokenHandler();
+            var key = Encoding.ASCII.GetBytes(jwtTokenSettings.Secret);
```
to-do leak reminder JwtSecurityTokenHandler()

enciney replied:
Normally in .net we have IDisposible class which is identifier that derived class will be disposed right away when the life cycle complete for this purpose we are using **using** keyword.
(https://dotnetcode.medium.com/understanding-idisposable-in-net-for-resource-management-638ec99a19f6)
Here JwtSecurityTokenHandler is not using un managed memory, this class is doing simple things like validating token and creating token, so we do not be worry about memory leak here

erenkorkmazemre replied:
Perfect, thanks for the garbage Collections Gods. But still I do not like it. Memory instance references will be destroy thats okay but we can also set just 1 instance and thats enough. Let me explain you with a draw when we are available.

enciney replied:
added as singleton

erenkorkmazemre commented on EternalBAND.DomainObjects/ApiContract/LoginInputContract.cs:
```diff
@@ -0,0 +1,9 @@
+    {
+        public string Username { get; set; }
+        public string Password { get; set; }
+        public bool RememberMe { get; set; } = false;
```
to-do cookie ui

enciney replied:
There are several options to hold information 

- Local Storage ( when browser tab closed it will be disposed)
- Session storage ( when browser windows closed it will be disposed)
- Cookies ( It will be persistent till user cleans the cookies manually ) 

So we can trust cookies since they are long time lived elements over browser. Since backend is trusting JWT token and looking for **Authorization** header in request, do not need to read or set Cookies. Only Front end will read/modify cookies.

Also we can hold tokens in cookie as well as hashed or BASE 64 encoded.

erenkorkmazemre replied:
- Local Storage: Not true
- Sessin storage: Half true
- Cookie: Half true

Pls look this [doc](https://stackoverflow.com/questions/19867599/what-is-the-difference-between-localstorage-sessionstorage-session-and-cookies) 
![image](https://github.com/user-attachments/assets/9c2d6ca2-c480-499d-9632-919486a926cc)

Also yes as we discuss before "to-do cookie ui" is written for the FE.  And yes we will encrypt our whole user infos etc or the cookies or storages etc.

erenkorkmazemre commented on EternalBAND.DomainObjects/ApiContract/LoginOutputContract.cs:
```diff
@@ -0,0 +1,8 @@
+    public class LoginOutputContract
+    {
+        public string Token { get; set; }
+        public string ReturnUrl { get; set; }
```
validation check 2 side, fe and be. Expire time decision

enciney replied:
We may integrate a flow right before the Request and basically add minimal validation rule. LoginOutputContract is output contract so no need for validation but for input contract like  "LoginInputContract" we can add this.

erenkorkmazemre replied:
- Yes and lets add the expire time in here. 
---
- Also yes validations for inputs. It will add for these.
![image](https://github.com/user-attachments/assets/145b8270-befe-4e27-98c0-df15e6eb8e1a)

enciney replied:
No need to hold expiration time of token as a separated property here, each endpoint may read this info from Authorization header on fly, also no need to.add manual expiry control , .net itself is managing it already.

I am planning to use generic types here so each validator can be usable according to contract data type 

IValidator<TType> and we can register as 

UserValidator as IValidator<User>

On the runtime we may instantiate related validator and use generic function call

erenkorkmazemre replied:
Yes it is true BE side dotnet is checking expire time with automatically, we will not validate from here for BE. But FE side we need this property, so we can keep it on our user claims or we can set directly login response. So the end of the day we need this one for the FE.

And yes as I mentioned in whatsapp, generic validator is more useful then others. Also FE will be use generic validators. Also annotations for help the validating process. In Spring Boot it is include,  maybe dotnet has also these kind of helping annotations.

erenkorkmazemre replied:
We decided. it is keeping automatic in claims(with a new JWT versions). So there is no need to set any expire time to somewhere.

enciney replied:
contract base  validation is added, fluent validator is being used as a validator library

enciney committed:
Signup is introduced as web api

Cookie is removed
Input contract validation is added
prod and dev appsettings conf. is created

enciney committed:
razor pages are removed
